Fix typed signatures parametric types

kdheepak opened the pull request:
Resolves #84. 

- [x] Fixes type signatures for parametric types
- [x] ~Removes `?` in type signature for functions with nameless arguments~

mortenpi commented:
Are you sure you're attacking the right place in the code at the moment? The error reported in #84, at least, is thrown in `abbreviations.jl`, in this loop specifically: https://github.com/JuliaDocs/DocStringExtensions.jl/blob/907520736dd0d1f61cd3796e8138bf48184f8b70/src/abbreviations.jl#L382-L391

`printmethod` seems to be already special-casing `UnionAll`s? I haven't look at it very closely though, so maybe I am just missing something about `printmethod`.

kdheepak commented:
I think there's more than one issue here. Here's what I'm planning to use to test:

```
import Base.Docs: meta, @var, DocStr, parsedoc

using DocStringExtensions
const DSE = DocStringExtensions
module M end

function find_function_name_symbol(e::Expr)
    if e.head != :call
        return find_function_name_symbol(e.args[1])
    else
        return e.args[1]
    end
end

function get_doc_ext_signature(sym, typesig)
    doc = Docs.DocStr(Core.svec(), nothing, Dict())
    buf = IOBuffer()
    doc.data = Dict(
        :binding => Docs.Binding(M, sym),
        :typesig => eval(typesig),
        :module => M,
    )
    DSE.format(DSE.TYPEDSIGNATURES, buf, doc)
    str = String(take!(buf))
    str = replace(str, "```julia" => "")
    str = replace(str, "```" => "")
    str = strip(str)
    return split(str, "\n")
end

exprs = [
            :(
                function no_arguments() end
             ),
            :(
                function one_argument_no_type(x) end
             ),
            :(
                function one_argument_typed(x::String) end
             ),
            :(
                function two_arguments_no_type(x, y) end
             ),
            :(
                function two_arguments_typed(x::String, y::Number) end
             ),
            :(
                function two_arguments_default(x::String, y::Number = 0) end
             ),
            :(
                function parametric_type_dispatch_on_type(::Type{T}, s::AbstractString) where T <: Number end
             ),
            :(
                function parametric_type_dispatch_on_value(::T, s::AbstractString) where T <: Number end
             ),
            :(
                function same_type(x::T, y::T) where T <: Number end
             ),
            :(
                function different_type_1(x::T, y::U) where T <: Number where U <: Number end
             ),
            :(
                function different_type_2(x::T, y::U) where {T<: Number, U <: Number} end
             ),
            :(
                function different_type_3(x::T, y::U) where {T, U <: Number} end
             ),
        ]

println()

for expr in exprs

    println("# `$(expr.args[1])`")
    println()

    docs_sig = Base.Docs.signature(expr)
    Base.eval(M, expr)
    f = find_function_name_symbol(expr)

    for m in methods(getproperty(M, f)).ms
        default_sig = m.sig
        @show default_sig
    end

    @show docs_sig

    docs_ext_sig = get_doc_ext_signature(f, docs_sig)
    @show docs_ext_sig
    println()
end
```

With the following exprs:

```
:(
    function parametric_type_dispatch_on_type(::Type{T}, s::AbstractString) where T <: Number end
 ),
:(
    function parametric_type_dispatch_on_value(::T, s::AbstractString) where T <: Number end
 ),
:(
    function different_type_1(x::T, y::U) where T <: Number where U <: Number end
 ),
:(
    function different_type_2(x::T, y::U) where {T<: Number, U <: Number} end
 ),
:(
    function different_type_3(x::T, y::U) where {T, U <: Number} end
 ),
```

I get 

```

# `parametric_type_dispatch_on_type(::Type{T}, s::AbstractString) where T <: Number`

default_sig = Tuple{typeof(Main.M.parametric_type_dispatch_on_type),Type{T},AbstractString} where T<:Number
docs_sig = :(Union{Tuple{Type{T}, AbstractString}, Tuple{T}} where T <: Number)
docs_ext_sig = SubString{String}["parametric_type_dispatch_on_type(?::T<:Number, s::T<:Number)"]

# `parametric_type_dispatch_on_value(::T, s::AbstractString) where T <: Number`

default_sig = Tuple{typeof(Main.M.parametric_type_dispatch_on_value),T,AbstractString} where T<:Number
docs_sig = :(Union{Tuple{T, AbstractString}, Tuple{T}} where T <: Number)
docs_ext_sig = SubString{String}["parametric_type_dispatch_on_value(?::T<:Number, s::T<:Number)"]

# `(different_type_1(x::T, y::U) where T <: Number) where U <: Number`

default_sig = Tuple{typeof(Main.M.different_type_1),T,U} where T<:Number where U<:Number
docs_sig = :((Union{Tuple{T, U}, Tuple{T}, Tuple{U}} where T <: Number) where U <: Number)
docs_ext_sig = SubString{String}["different_type_1(x::T<:Number, y::T<:Number)"]

# `different_type_2(x::T, y::U) where {T <: Number, U <: Number}`

default_sig = Tuple{typeof(Main.M.different_type_2),T,U} where U<:Number where T<:Number
docs_sig = :((Union{Tuple{T, U}, Tuple{U}, Tuple{T}} where U <: Number) where T <: Number)
docs_ext_sig = SubString{String}[""]

# `different_type_3(x::T, y::U) where {T, U <: Number}`

default_sig = Tuple{typeof(Main.M.different_type_3),T,U} where U<:Number where T
docs_sig = :((Union{Tuple{T, U}, Tuple{U}, Tuple{T}} where U <: Number) where T <: Any)
docs_ext_sig = SubString{String}["different_type_3(x::U<:Number, y::U<:Number)"]
```

So even the above commit is not doing the right thing. I'll come up with a more comprehensive list above first, and see what the values are for the default julia signature, the Base.Docs signature and the DocStringExtension generated signature and fix appropriately. I think the fix is straightforward once all the corner cases are identified.

kdheepak commented:
![Screen Shot 2020-05-12 at 9 28 09 AM](https://user-images.githubusercontent.com/1813121/81713307-ec2eff00-9432-11ea-9879-c32cb26eb774.png)

kdheepak commented:
![Screen Shot 2020-05-12 at 9 41 37 AM (2)](https://user-images.githubusercontent.com/1813121/81714897-cf93c680-9434-11ea-9a44-572346b7d2b5.png)

kdheepak commented:
This is ready to be merged.

mortenpi commented:
Thanks for your effort! I can see that indeed already fixes the error in #84. You are right though -- it seems like there are multiple related issues here.

First, just to make the review a bit easier for me (I quite unfamiliar with this codebase actually to be honest):

* Could you add docstrings to the new functions (`find_tuples` etc.), briefly explaining what they are doing?

* Also, the original code can be a bit tricky to follow. So, in case you had any aha! moments as you were puzzling it out, adding a few comments explaining what's happening here and there would be really valuable from a maintenance perspective.

Regarding the `where` syntax: it might be good if we could retain it. Currently, if you have a method like `foo(x::T, y::Vector{T}) where T <: Number = y` you seem to get a signature like this:

```
foo(x::T<:Number, y::Array{T<:Number,1}) -> Array{T<:Number,1}
```

whereas I think this would be more readable:

```
foo(x::T, y::Array{T,1}) where {T <: Number} -> Array{T,1}
```

Also, the first one is not valid Julia I believe.

Finally, it looks like there might be an issue if you have multiple methods:

```
julia> """
       $(TYPEDSIGNATURES)

       Bar.
       """
       foo(x::T) where T <: Number = x
foo

julia> """
       $(TYPEDSIGNATURES)

       Foo.
       """
       foo(x::T, y::Vector{T}) where T <: Number = y
foo

help?> foo
search: foo floor pointer_from_objref OverflowError RoundFromZero unsafe_copyto! functionloc

  foo(x::T<:Number) -> Number
  

  Bar.

  ──────────────────────────────────────────────────────────────────────────────────────────────────────

  foo(x::T<:Number) -> Array{T<:Number,1}
  foo(x::T<:Number, y::Array{T<:Number,1}) -> Array{T<:Number,1}
  

  Foo.
```

The second docstring shouldn't have two signatures. Cf. a non-parametric example:

```
julia> """
       $(TYPEDSIGNATURES)

       Foo.
       """
       foo(x::Int, y::Vector{Int}) = y
foo

julia> """
       $(TYPEDSIGNATURES)

       Foo.
       """
       foo(x::T, y::Vector{T}) where T <: Number = y

julia> """
       $(TYPEDSIGNATURES)

       Foo.
       """
       foo(x::Int) = x
foo

help?> foo
search: foo floor pointer_from_objref OverflowError RoundFromZero unsafe_copyto! functionloc

  foo(x::Int64, y::Array{Int64,1}) -> Array{Int64,1}
  

  Foo.

  ──────────────────────────────────────────────────────────────────────────────────────────────────────

  foo(x::Int64) -> Int64
  

  Foo.
```

kdheepak commented:
The following are notes on how DocStringExtensions generates type signatures.

Let's take the following function

```julia
module M

function two_arguments_default(x::String, y::Number = 0, z::Int = 0) end

end
```

There are three signatures for this function, one for each of its methods

1) `Tuple{typeof(Main.M.two_arguments_default),String}`
2) `Tuple{typeof(Main.M.two_arguments_default),String,Number}`
3) `Tuple{typeof(Main.M.two_arguments_default),String,Number,Int64}`

You can get these signatures using the Julia built in function `methods`:

```julia
for m in methods(two_arguments_default).ms
    @show m.sig
end
```

The built in Julia function `Base.Docs.signatures(expr)` generates a `Union` of these `Tuple`s, one for each type.

```julia
julia> signature(:(function two_arguments_default(x::String, y::Number = 0, z::Int = 0) end))

:(Union{Tuple{String}, Tuple{String, Number}, Tuple{String, Number, Int}})
```

DocStringExtensions generates docstrings for each of these methods.

```julia
doc.data = Dict(
    :binding => Docs.Binding(M, :two_arguments_default),
    :typesig => Union{Tuple{String}, Tuple{String, Number}, Tuple{String, Number, Int}},
    :module => M,
)
DSE.format(DSE.TYPEDSIGNATURES, buf, doc)
str = String(take!(buf))
@test all([occursin(s, str) for s in [
"two_arguments_default(x::String)",
"two_arguments_default(x::String, y::Number)",
"two_arguments_default(x::String, y::Number, z::Int64)",
]])
```

The entry point for the purposes of this discussion is the [`format`](https://github.com/kdheepak/DocStringExtensions.jl/blob/c494df899c6d69c388c08aa9a80d8e5390ee9bdc/src/abbreviations.jl#L366) function in DocStringExtensions.jl.

The various `method`s in `group` are the following:

```julia
Method[
    two_arguments_default(x::String) in Main.XDoc
    two_arguments_default(x::String, y::Number) in Main.XDoc
    two_arguments_default(x::String, y::Number, z::Int64) in Main.XDoc
]
```

To generate the docstring for each type, we call [`printmethod`](https://github.com/kdheepak/DocStringExtensions.jl/blob/c494df899c6d69c388c08aa9a80d8e5390ee9bdc/src/utilities.jl#L273)

For each `method` in the `group`, `arguments(method)` returns the argument symbols

```julia
args = [:x]
args = [:x, :y]
args = [:x, :y, :z]
```

When the correct `typesig` is passed, if it is a `Tuple` type, we can index the `.types` attribute with `i` to get the type of the `i`th argument.

We can also use this type signature and ask Julia's type inference to figure out the return types using the `Base.return_types` function.

For an example about generic functions, let's first take the following functions:

```julia
    """
    $(TYPEDSIGNATURES)

    Foo 1.
    """
    foo(x::Int) = x

    """
    $(TYPEDSIGNATURES)

    Foo 2.
    """
    foo(x::Int, y::Vector{Int}) = y
```

For function `foo`, there are two methods:

```julia
foo(x::Int64)
foo(x::Int64, y::Array{Int64,1}
```

The typesig generated by Julia is

```julia
Tuple{Int64}
Tuple{Int64,Array{Int64,1}}
```

`find_tuples(typesig)` returns a one element vector in each of these cases

```julia
DataType[Tuple{Int64}]
DataType[Tuple{Int64,Array{Int64,1}}]
```

This is what we saw earlier.

Let's take the following generic functions

```julia
    """
    $(TYPEDSIGNATURES)
    Bar 1.
    """
    bar(x::T) where T <: Number = x

    """
    $(TYPEDSIGNATURES)
    Bar 2.
    """
    bar(x::T, y::Vector{T}) where T <: Number = y
```

There are still two methods:

```
bar(x::T) where T<:Number
bar(x::T, y::Array{T,1}) where T<:Number
```

But the typesig generated by Julia is a `UnionAll`

```
Union{Tuple{T}, Tuple{T}} where T<:Number
Union{Tuple{T}, Tuple{T,Array{T,1}}} where T<:Number
```

I'm not really sure why this type signature is generated, it might be a bug. See https://github.com/JuliaLang/julia/issues/29437.

`find_tuples` returns a Vector of `Tuple`, where the last element is the full function signature.

```
find_tuples(Tuple{String,Number,Int64}) -> DataType[Tuple{String,Number,Int64}]
```

This is more relevant when the `typesig` is for a method with generic types.

`find_tuples(typesig)` can take a `UnionAll` type and return a vector all of the individual elements.
This is the result of `find_tuples` for each of the methods.

```julia
find_tuples(Union{Tuple{T}, Tuple{T}} where T<:Number) = Tuple{T<:Number}
find_tuples(Union{Tuple{T}, Tuple{T,Array{T,1}}} where T<:Number)[end] = Tuple{T<:Number,Array{T<:Number,1}}
```

We do this because the last element of the `UnionAll` is what we are interested in, as the type signature of the function.

The remaining parts of `printmethod` format the string appropriately.

kdheepak commented:
@mortenpi I simplified the implementation slightly. I also added some docstrings and some comments.

I've also posted a detailed explanation above of what is going on in the code. Hopefully that explains it a little more, so that you can review. And in the future people will be able to visit this PR for more information.

> Regarding the where syntax: it might be good if we could retain it. 

The `where` syntax would be nice. The part that I commented out in the codebase implements it, however I haven't implemented removing it from the existing arguments. I'll look into it, but maybe that can be a separate PR? I don't know if it is straightforward though. Consider the following function:

```julia
bar(x::T, y::Vector{<:Number}) where T <: AbstractString = y
```

This is the docstring that is generated.

```julia
bar(x::T<:AbstractString, y::Array{var"#s2",1} where var"#s2"<:Number) -> Array{var"#s2",1} where var"#s2"<:Number
```

What would you like this to be instead?

kdheepak commented:
Okay, I've made a change to support the `where` syntax only for explicit generic types. This is the docstrings generated by DocStringExtensions.jl now.

```julia
bar(x::T) where T<:Number -> Number
```

```julia
bar(x::T, y::Array{var"#s2",1} where var"#s2"<:Number) where T<:AbstractString -> Array{var"#s2",1} where var"#s2"<:Number
``` 

I've added a note in the code where I think it can be improved. I've not able to come up with a way to print out the correct generic type in all cases. I think that can be improved in a separate PR.

kdheepak commented:
@mortenpi what would it take to get this merged and in the next release?

mortenpi commented:
Sorry for the delay. It appears that there are still some issues to be ironed out -- the MWE from #87 errors still, although with a different message now:

```
julia> using DocStringExtensions

julia> abstract type Component end

julia> struct System end

julia> """
       $(TYPEDSIGNATURES)
       ...
       """
       function get_components(
           ::Type{T},           sys::System,
           filter_func::Union{Nothing, Function} = nothing,
       ) where {T <: Component}
           # TODO: Verify that return type annotation is not required           #return IS.get_components(T, sys.data, filter_func)
       end
get_components

help?> get_components
search: get_components

ERROR: MethodError: no method matching type_to_string(::Type{Union{Nothing, Function}})
Closest candidates are:
  type_to_string(::TypeVar) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/utilities.jl:292
  type_to_string(::DataType) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/utilities.jl:283
  type_to_string(::UnionAll) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/utilities.jl:304Stacktrace:
 [1] printmethod(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Base.Docs.Binding, ::Function, ::Method, ::Type{T} where T) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/utilities.jl:334
 [2] format(::DocStringExtensions.TypedMethodSignatures, ::Base.GenericIOBuffer{Array{UInt8,1}}, ::Base.Docs.DocStr) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/abbreviations.jl:387
 [3] formatdoc(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Base.Docs.DocStr, ::DocStringExtensions.TypedMethodSignatures) at /home/mortenpi/Julia/JuliaDocs/DocStringExtensions/src/abbreviations.jl:25
 [4] formatdoc(::Base.Docs.DocStr) at /home/mortenpi/Julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/docview.jl:60
 [5] parsedoc(::Base.Docs.DocStr) at /home/mortenpi/Julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/docview.jl:68
 [6] iterate at ./generator.jl:47 [inlined]
 [7] _collect(::Array{Base.Docs.DocStr,1}, ::Base.Generator{Array{Base.Docs.DocStr,1},typeof(Base.Docs.parsedoc)}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./array.jl:678
 [8] collect_similar at ./array.jl:607 [inlined]
 [9] map at ./abstractarray.jl:2072 [inlined]
 [10] doc(::Base.Docs.Binding, ::Type{T} where T) at /home/mortenpi/Julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/docview.jl:117
 [11] doc(::Base.Docs.Binding) at /home/mortenpi/Julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/docview.jl:85
 [12] top-level scope at /home/mortenpi/Julia/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/docview.jl:287
```

To allay any doubts, we'll definitely have this in the next release, which we can tag as soon as this is merged.

mortenpi commented on src/utilities.jl:
```diff
@@ -222,6 +222,88 @@ function printmethod(buffer::IOBuffer, binding::Docs.Binding, func, method::Meth
+
+Find the tuples that represent all the types of all the method arguments and return a DataType Vector
+
+Tuple{String,Number,Int} -> DataType[Tuple{String,Number,Int}]
+
+Union{Tuple{Int64}, Tuple{U}, Tuple{T}, Tuple{Int64,T}, Tuple{Int64,T,U}} where U where
+T -> DataType[Tuple{Int64}, Tuple{U}, Tuple{T}, Tuple{Int64,T}, Tuple{Int64,T,U}]
```
````suggestion
```jldoctest; setup = :(using DocStringExtensions)
julia> DocStringExtensions.find_tuples(Tuple{String,Number,Int})
1-element Array{DataType,1}:
 Tuple{String,Number,Int64}

julia> DocStringExtensions.find_tuples(Tuple{T} where T <: Integer)
1-element Array{DataType,1}:
 Tuple{T<:Integer}

julia> s = Union{
         Tuple{Int64},
         Tuple{U},
         Tuple{T},
         Tuple{Int64,T},
         Tuple{Int64,T,U}
       } where U where T;

julia> DocStringExtensions.find_tuples(s)
5-element Array{DataType,1}:
 Tuple{Int64}
 Tuple{U}
 Tuple{T}
 Tuple{Int64,T}
 Tuple{Int64,T,U}
```
````

mortenpi commented on src/utilities.jl:
```diff
@@ -222,6 +222,88 @@ function printmethod(buffer::IOBuffer, binding::Docs.Binding, func, method::Meth
+    end
+end
+
+"""
+$(:SIGNATURES)
+
+This function takes any type signature that returns an Vector of `TypeVar`
+"""
+function find_vars(typesig)
+    if typesig isa UnionAll
+        return [find_vars(typesig.body)...]
+    elseif typesig isa TypeVar
+        return [typesig]
+    else
+        return []
+    end
+end
+
```
I don't think `find_vars` is used anywhere?

```suggestion
```

mortenpi commented on src/utilities.jl:
```diff
@@ -262,12 +343,11 @@ function printmethod(buffer::IOBuffer, binding::Docs.Binding, func, method::Meth
-    else
-        t = typesig
+    for t in reverse(where_syntax)
+        s = typevar_to_string(t)
```
It seems that this is the only place where `typevar_to_string` is used. Would it make sense to do this instead and get rid of that function?

```suggestion
        s = "where $t" # t is a TypeVar
```

mortenpi commented on src/utilities.jl:
```diff
@@ -231,7 +313,6 @@ simplifications include:
   * no `TypeVar`s;
   * no types;
   * no keyword default values;
-  * `?` printed where `#unused#` arguments are found.
```
This looks like an unrelated change? It would also apply to `SIGNATURES`? It would be good to discuss this in a separate PR. Although I am not sure if we want this:

```
julia> h(x, ::Int, z) = x ;

# master
julia> printmethod(Docs.Binding(Main, :h), h, first(methods(h)))
"h(x, ?, z)"

# vs this PR
julia> printmethod(Docs.Binding(Main, :h), h, first(methods(h)))
"h(x, , z)"
```

MichaelHatherly replied:
Using underscores might be the least confusing syntax here since it's already valid syntax:

```
julia> h(x, _, z) = x
h (generic function with 1 method)
```

I recall `?` being added prior to the official `_` syntax being introduced.

Either way, separate PR seems appropriate.

kdheepak replied:
I changed it to `_` and added a test for the `h(x, _, z)` function. Resolved in f75bad9, 87df61e and 6d9f77a.

mortenpi replied:
Changing to underscores is a good idea and the change LGTM. But to keep unrelated changes in separate commits, could you open a separate PR for this? I think it's only 2 lines + the test to move to move over?

kdheepak replied:
Sounds good! I will do that.

kdheepak commented:
@mortenpi I've fixed the MWE from that issue and added test for it as well. Ironically, that's the issue that had me make this PR 😅. See 2080a7b.

There is a caveat though. The `where T<:Type` only works for typevars at the moment. You'll see in the test that `::Type{T<:Number}, x::String` is printed instead of `::Type{T}, x::String, ...) where T <: Number`. I think making this work in the general case is not straightforward, since you have to recursively walk through a type to figure out if there's a `TypeVar` in there or not, but it can be done. The alternative is to remove the `where` clause and have it always compress the values into the function argument.

Thoughts?

kdheepak commented:
I've removed the commits related to the `_` to `?`. I'll submit a separate PR as soon as this is merged.

mortenpi commented:
Sorry, I manage to keep finding ways of breaking this :sweat_smile: For some reason `Vector` disappears here:

```
julia> """
       $(TYPEDSIGNATURES)
       """
       foo(x::Vector{T}) where {T <: Integer} = 0

help?> foo

  foo(x::T) where T<:Integer
```

And also type inference for the return type is also gone.

kdheepak commented:
> For some reason `Vector` disappears

oops, I'll look into it, sorry about about. Better to catch these now than later!

kdheepak commented:
This issue should be fixed now. It'd be useful to run a more comprehensive test suite at some point.

jd-lara commented:
bump @kdheepak and @mortenpi about this PR

mortenpi commented:
Sorry again for not reviewing earlier. I did come across this case:

```julia
"""
$(TYPEDSIGNATURES)
"""
bar(x::Union{T,Nothing}, y::T = zero(T)) where {T <: Number} = x
```

leads to

```
bar(x::Union{Nothing, T<:Number}) -> Union{Nothing, Number}
bar(x::Union{Nothing, T<:Number}, y::T) where T<:Number -> Union{Nothing, T<:Number}
```

1. The return type is slightly different between the two. Is that because the type inference returns a slightly different result for these two methods?
2. The second return type should be either `Union{Nothing, Number}` or `Union{Nothing, T} where {T<:Number}` I think?
3. In the second case it puts the redundant `<:Number` in the `x::Union{...}`.

@kdheepak if you have time to look into this, it would be awesome. But I also feel that we can spend an infinite amount of time iterating on this, so I would be happy to ship this as is too (with some known cosmetic bugs). It doesn't seem to crash anymore, which is a massive improvement over the status quo.

Just as a side note: a couple of ideas for future improvement (definitely beyond the scope of this PR):

* Apply some simplifications for common cases (e.g. `Array{T,1}` -> `Vector{T}`).
* If the signature is longer than some number of characters (92?), split it over multiple lines (probably each variable and `where` typevar on separate lines)

kdheepak commented:
Thanks for finding this!

> The return type is slightly different between the two. Is that because the type inference returns a slightly different result for these two methods?

Type inference is returning that. In my version of julia this is what I get:

```
println(VERSION)
println(@doc(XDoc.bar))
```

````
1.6.0-DEV.138
```julia
bar(x::Union{Nothing, T<:Number}) -> Union{Nothing, Number}
bar(x::Union{Nothing, T<:Number}, y::T) where T<:Number -> Union{Nothing, T<:Number}
```
````

> The second return type should be either Union{Nothing, Number} or Union{Nothing, T} where {T<:Number} I think?

Again this is just return types from Julia directly directly string interpolated. I agree that we should convert these return types to nicer syntax, and I think I agree that we should do it in a separate PR.

> In the second case it puts the redundant <:Number in the x::Union{...}.

I'll look into this now!

kdheepak commented:
> In the second case it puts the redundant <:Number in the x::Union{...}.

What is happening here in this case is hard to fix for me. The current implementation has a comment for this as well, I only added `where T` for `TypeVar` but not `Union` types. In this example that you identified, the first argument is a `Union` type and the second argument is a `TypeVar`. 

I think the best path forward is to not implement the `where T` syntax. It is currently very fragile and definitely does not take into consideration all the ways Julia types can be defined and if this is merged as is it'll be a annoyance for users because their documentation will be inconsistent.

I think it is better to have the generic types in the argument itself. This way, it is consistent across the documentation. The implementation is easier and all the corner cases can be covered. The downside is that it is more verbose and not idiomatic Julia. But that can be fixed in a separate PR in the future along with the other type string changes.

jd-lara commented:
bump to @mortenpi about this PR

mortenpi commented on test/tests.jl:
````diff
@@ -212,6 +223,146 @@ end
+            @test occursin("\n```julia\n", str)
+            if VERSION > v"1.3.0"
+                @test occursin("\nk_6(x::Array{T<:Number,1}) -> Array{T<:Number,1}\n", str)
+            end
````
Why are we version-guarding these tests? Ditto for the tests below.

kdheepak replied:
I've added a version guard for the only function that requires it in the test. Resolved in 403028e.

mortenpi commented on test/tests.jl:
````diff
@@ -212,6 +223,146 @@ end
+            if VERSION > v"1.3.0"
+                @test_broken occursin("\nk_10(x::T) -> Any\n", str)
+            end
+            @test_broken occursin("\n```\n", str)
````
What is this one testing actually?

kdheepak replied:
This test at the moment is probably a repeat of another function. But it will be useful to have when the `where T` syntax is introduced.

kdheepak commented:
I've made the changes you've requested! Thank you taking the time to review again!

mortenpi commented:
Lovely! Thanks @kdheepak for all the work here and sorry it took so long to get this merged.

Will you be doing the `_` PR as well?

kdheepak commented:
Thanks for merging! And thanks for being so patient and finding all the corner cases you did! I'm sure more will come up in the future. Feel free to ping me whenever that happens.

I'll submit the `_` PR now.
